File: testSuite/test-transfer-functions-axion.pl
#!/usr/bin/env perl
use strict;
use warnings;
use lib $ENV{'GALACTICUS_EXEC_PATH'         }."/perl";
use lib $ENV{'GALACTICUS_ANALYSIS_PERL_PATH'}."/perl";
use PDL;
use PDL::NiceSlice;
use PDL::IO::HDF5;
use GnuPlot::PrettyPlots;
use GnuPlot::LaTeX;

my @models =
    (
     {
	 label                   => "CDM"                      ,
	 plotLabel               => "CDM"                      ,
	 testPeaks               => "no"
     },
     {
	 label                   => "AxionCAMB"                ,
	 plotLabel               => "AxionCAMB"                ,
	 color                   => "redYellow"                ,
	 testPeaks               => "no"
     },
     {
	 label                   => "AxionMurgia2017"          ,
	 plotLabel               => "Murgia et al. (2017)"     ,
	 color                   => "peachPuff"                ,
	 testPeaks               => "no"                       ,
	 toleranceCutOff         => 0.10
     },
     {
	 label                   => "AxionHu2000"              ,
	 plotLabel               => "Hu et al. (2000)"         ,
	 color                   => "mediumSeaGreen"           ,
	 testPeaks               => "yes"                      ,
	 toleranceCutOff         => 0.160                      ,
	 tolerancePeakWavenumber => 0.090                      ,
	 tolerancePeakAmplitude  => 0.800
     },
     {
	 label                   => "AxionPassaglia2022"       ,
	 plotLabel               => "Passaglia \\\\& Hu (2022)",
	 color                   => "cornflowerBlue"           ,
	 testPeaks               => "yes"                      ,
	 toleranceCutOff         => 0.025                      ,
	 tolerancePeakWavenumber => 0.015                      ,
	 tolerancePeakAmplitude  => 0.150
     }
    );
my $transferFunctions;
foreach my $model ( @models ) {
    # Run models.
    system("mkdir -p outputs; cd ..; ./Galacticus.exe testSuite/parameters/powerSpectrum".$model->{'label'}.".xml");
    unless ( $? == 0 ) {
    	print "FAIL: failed to run model 'powerSpectrum".$model->{'label'}.".xml'\n";
    	exit;
    }
    # Read data.
    my $galacticus = new PDL::IO::HDF5("outputs/powerSpectrum".$model->{'label'}.".hdf5");
    my $output     = $galacticus->group('Outputs/Output1');
    $transferFunctions->{$model->{'label'}}->{$_} = $output->dataset($_)->get()
	foreach ( "wavenumber", "transferFunction" );
    # Normalize the transfer function.
    $transferFunctions->{$model->{'label'}}->{'transferFunction'} /=     $transferFunctions->{$model->{'label'}}->{'transferFunction'}->((0));
    $transferFunctions->{$model->{'label'}}->{'transferFunction'} .= abs($transferFunctions->{$model->{'label'}}->{'transferFunction'}      );
    $transferFunctions->{$model->{'label'}}->{'transferFunction'} /=     $transferFunctions->{         'CDM'   }->{'transferFunction'}
        unless ( $model->{'label'} eq "CDM" );
}

# Create a plot of the window function.
my $plot;
my $gnuPlot;
my $plotFileTeX = "outputs/transferFunctionsAxion.tex";
open($gnuPlot,"|gnuplot");
print $gnuPlot "set terminal cairolatex pdf standalone color lw 2 size 3.5in,3.5in\n";
print $gnuPlot "set output '".$plotFileTeX."'\n";
print $gnuPlot "set lmargin screen 0.20\n";
print $gnuPlot "set rmargin screen 0.95\n";
print $gnuPlot "set bmargin screen 0.15\n";
print $gnuPlot "set tmargin screen 0.95\n";
print $gnuPlot "set key spacing 1.2\n";
print $gnuPlot "set key at screen 0.9,0.93\n";
print $gnuPlot "set xlabel '\$ k \$ Mpc\$^{-1}\$'\n";
print $gnuPlot "set ylabel '\$ (k/10\\mathrm{Mpc}^{-1})^6 |T(k)| \$\n";
print $gnuPlot "set logscale y\n";
print $gnuPlot "set mytics 10\n";
print $gnuPlot "set format y '\$10^{\%L}\$'\n";
print $gnuPlot "set xrange [3.0e+0:15.0]\n";
print $gnuPlot "set yrange [1.0e-3:3.0e-1]\n";
print $gnuPlot "set pointsize 1.0\n";
foreach my $model ( @models ) {
    next
	if ( $model->{'label'} eq "CDM" );
    &GnuPlot::PrettyPlots::Prepare_Dataset(
	    \$plot                                                           ,
	    +  $transferFunctions->{$model->{'label'}}->{'wavenumber'      } ,
	    +  $transferFunctions->{$model->{'label'}}->{'transferFunction'}
	    *(
	      +$transferFunctions->{$model->{'label'}}->{'wavenumber'      }
	      /10.0
	    )**6,
	    style  => "line"                                                 ,
	    weight => [2,1]                                                  ,
	    color  => $GnuPlot::PrettyPlots::colorPairs{$model->{'color'}}   ,
	    title  => $model->{'plotLabel'}
	);
}
&GnuPlot::PrettyPlots::Plot_Datasets($gnuPlot,\$plot);
close($gnuPlot);
&GnuPlot::LaTeX::GnuPlot2PDF($plotFileTeX);

# Perform tests.
foreach my $model ( @models ) {
    next
	if ( $model->{'label'} eq "CDM" );
    # Find locations of peaks.
    @{$transferFunctions->{$model->{'label'}}->{'peaks'}} = ();
    for(my $i=1;$i<nelem($transferFunctions->{$model->{'label'}}->{'wavenumber'})-1;++$i) {
	push(
	    @{$transferFunctions->{$model->{'label'}}->{'peaks'}},
	    {
		wavenumber       => $transferFunctions->{$model->{'label'}}->{'wavenumber'      }->(($i)),
		transferFunction => $transferFunctions->{$model->{'label'}}->{'transferFunction'}->(($i))
	    }
	    )
	    if (
		$transferFunctions->{$model->{'label'}}->{'transferFunction'}->(($i)) > $transferFunctions->{$model->{'label'}}->{'transferFunction'}->(($i-1))
		&&
		$transferFunctions->{$model->{'label'}}->{'transferFunction'}->(($i)) > $transferFunctions->{$model->{'label'}}->{'transferFunction'}->(($i+1))
	    );
    }
    # Test amplitudes at ~4 Mpc⁻¹ against AxionCAMB.
    if ( exists($model->{'toleranceCutOff'}) ) {
	my $wavenumberTarget = pdl 4.0;
	my $indexTarget      = minimum_ind(abs($transferFunctions->{$model->{'label'}}->{'wavenumber'}-$wavenumberTarget));
	my $error =
	    abs(
		+$transferFunctions->{$model->{'label'    }}->{'transferFunction'}->(($indexTarget))
		-$transferFunctions->{         'AxionCAMB' }->{'transferFunction'}->(($indexTarget))
	    )
	    /    $transferFunctions->{         'AxionCAMB' }->{'transferFunction'}->(($indexTarget));
	my $errorPercent = sprintf("%4.1f%%",100.0*$error);
	my $status       = $error > $model->{'toleranceCutOff'} ? "FAIL" : "SUCCESS";
	print $model->{'label'}." T(k=4 Mpc⁻¹): ".$errorPercent." (".$status.")\n";
    }
    # Test locations and amplitudes of peaks.
    if ( exists($model->{'tolerancePeakWavenumber'}) ) {
	my @subscripts = ("₀","₁","₂","₃","₄","₅","₆","₇","₈","₉");
	foreach(my $peak=0;$peak<2;++$peak) {
	    my $errorWavenumber =
		abs(
		    +$transferFunctions->{$model->{'label'    }}->{'peaks'}->[$peak]->{'wavenumber'      }
		    -$transferFunctions->{         'AxionCAMB' }->{'peaks'}->[$peak]->{'wavenumber'      }
		)
		/    $transferFunctions->{         'AxionCAMB' }->{'peaks'}->[$peak]->{'wavenumber'      };
	    my $errorAmplitude =
		abs(
		    +$transferFunctions->{$model->{'label'    }}->{'peaks'}->[$peak]->{'transferFunction'}
		    -$transferFunctions->{         'AxionCAMB' }->{'peaks'}->[$peak]->{'transferFunction'}
		)
		/    $transferFunctions->{         'AxionCAMB' }->{'peaks'}->[$peak]->{'transferFunction'};
	    my $errorWavenumberPercent = sprintf("%4.1f%%",100.0*$errorWavenumber);
	    my $errorAmplitudePercent  = sprintf("%4.1f%%",100.0*$errorAmplitude );
	    my $statusWavenumber       = $errorWavenumber > $model->{'tolerancePeakWavenumber'} ? "FAIL" : "SUCCESS";
	    my $statusAmplitude        = $errorAmplitude  > $model->{'tolerancePeakAmplitude' } ? "FAIL" : "SUCCESS";
	    print $model->{'label'}." k".$subscripts[$peak+1].": ".$errorWavenumberPercent." (".$statusWavenumber.")\n";
	    print $model->{'label'}." T".$subscripts[$peak+1].": ".$errorAmplitudePercent ." (".$statusAmplitude .")\n";
	}
    }
}

exit;

File: testSuite/test-transfer-functions-axion.py
#!/usr/bin/env python3
import subprocess
import sys
import h5py
import numpy as np

# Test axion transfer functions.
# Andrew Benson (ported to Python)

models = [
    {"label": "CDM",               "testPeaks": "no"},
    {"label": "AxionCAMB",         "testPeaks": "no"},
    {"label": "AxionMurgia2017",   "testPeaks": "no",  "toleranceCutOff": 0.10},
    {"label": "AxionHu2000",       "testPeaks": "yes", "toleranceCutOff": 0.160, "tolerancePeakWavenumber": 0.090, "tolerancePeakAmplitude": 0.802},
    {"label": "AxionPassaglia2022","testPeaks": "yes", "toleranceCutOff": 0.025, "tolerancePeakWavenumber": 0.017, "tolerancePeakAmplitude": 0.165},
]

# Ensure output directory exists.
subprocess.run("mkdir -p outputs", shell=True)

transferFunctions = {}

for model in models:
    # Run models.
    status = subprocess.run(
        f"cd ..; ./Galacticus.exe testSuite/parameters/powerSpectrum{model['label']}.xml",
        shell=True
    )
    if status.returncode != 0:
        print(f"FAIL: failed to run model 'powerSpectrum{model['label']}.xml'")
        sys.exit(0)

    # Read data.
    with h5py.File(f"outputs/powerSpectrum{model['label']}.hdf5", "r") as galacticus:
        output = galacticus["Outputs/Output1"]
        wavenumber       = output["wavenumber"][:]
        transferFunction = output["transferFunction"][:]

    # Normalize the transfer function.
    transferFunction /= transferFunction[0]
    transferFunction  = np.abs(transferFunction)

    transferFunctions[model["label"]] = {
        "wavenumber":       wavenumber,
        "transferFunction": transferFunction,
    }

# Divide by CDM transfer function for non-CDM models.
for model in models:
    if model["label"] != "CDM":
        transferFunctions[model["label"]]["transferFunction"] /= transferFunctions["CDM"]["transferFunction"]

# Perform tests.
for model in models:
    if model["label"] == "CDM":
        continue

    label  = model["label"]
    tf     = transferFunctions[label]["transferFunction"]
    wn     = transferFunctions[label]["wavenumber"]

    # Find locations of peaks.
    peaks = []
    for i in range(1, len(wn) - 1):
        if tf[i] > tf[i-1] and tf[i] > tf[i+1]:
            peaks.append({"wavenumber": wn[i], "transferFunction": tf[i]})
    transferFunctions[label]["peaks"] = peaks

    # Test amplitudes at ~4 Mpc^-1 against AxionCAMB.
    if "toleranceCutOff" in model:
        wavenumberTarget = 4.0
        indexTarget      = np.argmin(np.abs(wn - wavenumberTarget))
        error            = abs(tf[indexTarget] - transferFunctions["AxionCAMB"]["transferFunction"][indexTarget]) / transferFunctions["AxionCAMB"]["transferFunction"][indexTarget]
        errorPercent     = f"{100.0 * error:4.1f}%"
        status_str       = "FAIL" if error > model["toleranceCutOff"] else "SUCCESS"
        print(f"{label} T(k=4 Mpc\u207b\u00b9): {errorPercent} ({status_str})")

    # Test locations and amplitudes of peaks.
    if "tolerancePeakWavenumber" in model:
        subscripts = ["\u2080", "\u2081", "\u2082", "\u2083", "\u2084", "\u2085", "\u2086", "\u2087", "\u2088", "\u2089"]
        for peak in range(2):
            axionCAMBPeaks = transferFunctions["AxionCAMB"]["peaks"]
            modelPeaks     = transferFunctions[label]["peaks"]
            errorWavenumber = (
                abs(modelPeaks[peak]["wavenumber"] - axionCAMBPeaks[peak]["wavenumber"])
                / axionCAMBPeaks[peak]["wavenumber"]
            )
            errorAmplitude = (
                abs(modelPeaks[peak]["transferFunction"] - axionCAMBPeaks[peak]["transferFunction"])
                / axionCAMBPeaks[peak]["transferFunction"]
            )
            errorWavenumberPercent = f"{100.0 * errorWavenumber:4.1f}%"
            errorAmplitudePercent  = f"{100.0 * errorAmplitude:4.1f}%"
            statusWavenumber = "FAIL" if errorWavenumber > model["tolerancePeakWavenumber"] else "SUCCESS"
            statusAmplitude  = "FAIL" if errorAmplitude  > model["tolerancePeakAmplitude"]  else "SUCCESS"
            print(f"{label} k{subscripts[peak+1]}: {errorWavenumberPercent} ({statusWavenumber})")
            print(f"{label} T{subscripts[peak+1]}: {errorAmplitudePercent} ({statusAmplitude})")
